In [77]:
from sqlalchemy import create_engine, inspect
from sqlalchemy.engine import reflection
import pandas as pd

engine = create_engine("sqlite:///sqlite-sakila.db")
inspector = inspect(engine)
inspector.get_table_names()

['actor',
 'address',
 'category',
 'city',
 'country',
 'customer',
 'film',
 'film_actor',
 'film_category',
 'film_text',
 'inventory',
 'language',
 'payment',
 'rental',
 'staff',
 'store']

In [57]:
"""
initialize engine
"""
for i in inspector.get_columns('film'):
    if i["name"] in ["title", "description",	"special_features"]:
        print(i["name"], i["type"])

title VARCHAR(255)
description TEXT
special_features VARCHAR(100)


In [7]:
pd.read_sql_query('''
SELECT  title,
        description,
        special_features
FROM film
LIMIT 5''', engine)

title                                        description  \
0  ACADEMY DINOSAUR  A Epic Drama of a Feminist And a Mad Scientist...   
1    ACE GOLDFINGER  A Astounding Epistle of a Database Administrat...   
2  ADAPTATION HOLES  A Astounding Reflection of a Lumberjack And a ...   
3  AFFAIR PREJUDICE  A Fanciful Documentary of a Frisbee And a Lumb...   
4       AFRICAN EGG  A Fast-Paced Documentary of a Pastry Chef And ...   

                   special_features  
0  Deleted Scenes,Behind the Scenes  
1           Trailers,Deleted Scenes  
2           Trailers,Deleted Scenes  
3    Commentaries,Behind the Scenes  
4                    Deleted Scenes

ใน PostgreSQL Database สามารถดูข้อมูลจาก query นี้ได้

```
SELECT column_name, data_type
FROM INFORMATION_SCHEMA.COLUMNS
WHERE column_name in ('title', 'description', 'special_features') AND table_name = 'film';
```

แต่ถ้าเป็น sqlite ทำไม่ได้ ต้องเขียนเอง ใช้ inspector object


In [64]:
from sqlalchemy import create_engine, inspect
from sqlalchemy.engine import reflection
import pandas as pd

engine = create_engine("sqlite:///sqlite-sakila.db")
inspector = inspect(engine)

for i in inspector.get_columns('film'): # inspector.get_columns('film') เป็น list โดยที่ข้างในแต่ะตัวเป็น dict
    if i["name"] in ["title", "description",	"special_features"]:
        print(i["name"], i["type"])

title VARCHAR(255)
description TEXT
special_features VARCHAR(100)


### จริงๆ special_features ต้องเป็น array แต่ใน sqlite ไม่มี array

### Getting information about your database
As we saw in the video, PostgreSQL has a system database called `INFORMATION_SCHEMA` that allows us to extract information about objects, including tables, in our database.

In this exercise we will look at how to query the `tables` table of the `INFORMATION_SCHEMA` database to discover information about tables in the DVD Rentals database including the name, type, schema, and catalog of all tables and views and then how to use the results to get additional information about columns in our tables.

* Select all columns from the `INFORMATION_SCHEMA.TABLES` system database. Limit results that have a public `table_schema`.

```
SELECT * 
FROM INFORMATION_SCHEMA.TABLES
 -- Filter by schema
 WHERE table_schema = 'public';
```

```
table_catalog	table_schema	table_name	table_type	self_referencing_column_name	reference_generation	user_defined_type_catalog	user_defined_type_schema	user_defined_type_name	is_insertable_into	is_typed	commit_action
pgdata	public	actor	BASE TABLE	null	null	null	null	null	YES	NO	null
pgdata	public	category	BASE TABLE	null	null	null	null	null	YES	NO	null
```

* Select all columns from the `INFORMATION_SCHEMA.COLUMNS` system database. Limit by `table_name` to `actor`

```
 SELECT * 
 FROM INFORMATION_SCHEMA.COLUMNS 
 WHERE table_name = 'actor';
 
 
 
table_catalog	table_schema	table_name	column_name	ordinal_position	column_default	is_nullable	data_type	character_maximum_length	character_octet_length	numeric_precision	numeric_precision_radix	numeric_scale	datetime_precision	interval_type	interval_precision	character_set_catalog	character_set_schema	character_set_name	collation_catalog	collation_schema	collation_name	domain_catalog	domain_schema	domain_name	udt_catalog	udt_schema	udt_name	scope_catalog	scope_schema	scope_name	maximum_cardinality	dtd_identifier	is_self_referencing	is_identity	identity_generation	identity_start	identity_increment	identity_maximum	identity_minimum	identity_cycle	is_generated	generation_expression	is_updatable
pgdata	public	actor	actor_id	1	null	YES	smallint	null	null	16	2	0	null	null	null	null	null	null	null	null	null	null	null	null	pgdata	pg_catalog	int2	null	null	null	null	1	NO	NO	null	null	null	null	null	NO	NEVER	null	YES
pgdata	public	actor	first_name	2	null	YES	character varying	45	180	null	null	null	null	null	null	null	null	null	null	null	null	null	null	null	pgdata	pg_catalog	varchar	null	null	null	null	2	NO	NO	null	null	null	null	null	NO	NEVER	null	YES
pgdata	public	actor	last_name	3	null	YES	character varying	45	180	null	null	null	null	null	null	null	null	null	null	null	null	null	null	null	pgdata	pg_catalog	varchar	null	null	null	null	3	NO	NO	null	null	null	null	null	NO	NEVER	null	YES

```

### Determining data types
The columns table of the `INFORMATION_SCHEMA` database also allows us to extract information about the data types of columns in a table. We can extract information like the character or string length of a `CHAR` or `VARCHAR` column or the precision of a `DECIMAL` or `NUMERIC` floating point type.

Using the techniques you learned in the lesson, let's explore the customer table of our DVD Rental database.

* Select the column name and data type from the `INFORMATION_SCHEMA.COLUMNS` system database.
* Limit results to only include the `customer` table.

```
SELECT column_name, data_type
FROM INFORMATION_SCHEMA.COLUMNS
WHERE table_name = 'customer';
```

In [67]:
for i in inspector.get_columns('customer'): # inspector.get_columns('film') เป็น list โดยที่ข้างในแต่ะตัวเป็น dict
    print(i["name"], i["type"])

customer_id INTEGER
store_id INTEGER
first_name VARCHAR(45)
last_name VARCHAR(45)
email VARCHAR(50)
address_id INTEGER
active CHAR(1)
create_date TIMESTAMP
last_update TIMESTAMP


# TIMESTAMP datatype

In [75]:
pd.read_sql('''
SELECT payment_date
FROM payment LIMIT 2;''', engine)

payment_date
0  2005-05-25 11:30:37.000
1  2005-05-28 10:35:23.000

In [74]:
for i in inspector.get_columns('payment'): # inspector.get_columns('film') เป็น list โดยที่ข้างในแต่ะตัวเป็น dict
    if i["name"] == "payment_date":
        print(i["name"], i["type"])

payment_date TIMESTAMP


# DATE and TIME datatypes
* DATE: 2005-10-31
* TIME: 01:05:30.9.3027+00

In [94]:
pd.read_sql('''
SELECT payment_date AS timestamp, 
       DATE(payment_date) AS date,
       TIME(payment_date) AS time
FROM payment LIMIT 2;''', engine)

timestamp        date      time
0  2005-05-25 11:30:37.000  2005-05-25  11:30:37
1  2005-05-28 10:35:23.000  2005-05-28  10:35:23

# INTERVAL datatypes

```
SELECT rental_date + INTERVAL '3 days' AS expected_return
FROM rental
```

ใน SQLite ไม่มี `INTERVAL` ต้องใช้แบบด้านล่าง

In [104]:
pd.read_sql_query('''
SELECT DATE(rental_date), DATE(rental_date, '+3 days') AS expected_return
FROM rental''', engine).head()

DATE(rental_date) expected_return
0        2005-05-24      2005-05-27
1        2005-05-24      2005-05-27
2        2005-05-24      2005-05-27
3        2005-05-24      2005-05-27
4        2005-05-24      2005-05-27

### Interval data types

`INTERVAL` data types provide you with a very useful tool for performing arithmetic on date and time data types. For example, let's say our rental policy requires a DVD to be returned within 3 days. We can calculate the `expected_return_date` for a given DVD rental by adding an `INTERVAL` of 3 days to the `rental_date` from the rental table. We can then compare this result to the actual `return_date` to determine if the DVD was returned late.

* Select the rental date and return date from the `rental` table.
* Add an `INTERVAL` of 3 days to the `rental_date` to calculate the expected return date`.

```
SELECT rental_date,
	   return_date,
 	   rental_date + INTERVAL '3 days' AS expected_return_date
FROM rental;
```

In [103]:
pd.read_sql_query('''
SELECT rental_date,
       return_date,
       DATE(rental_date, '+3 days') AS expected_return_date
FROM rental; ''', engine)

rental_date              return_date expected_return_date
0      2005-05-24 22:53:30.000  2005-05-26 22:04:30.000           2005-05-27
1      2005-05-24 22:54:33.000  2005-05-28 19:40:33.000           2005-05-27
2      2005-05-24 23:03:39.000  2005-06-01 22:12:39.000           2005-05-27
3      2005-05-24 23:04:41.000  2005-06-03 01:43:41.000           2005-05-27
4      2005-05-24 23:05:21.000  2005-06-02 04:33:21.000           2005-05-27
...                        ...                      ...                  ...
16039  2005-08-23 22:25:26.000  2005-08-25 23:54:26.000           2005-08-26
16040  2005-08-23 22:26:47.000  2005-08-27 18:02:47.000           2005-08-26
16041  2005-08-23 22:42:48.000  2005-08-25 02:48:48.000           2005-08-26
16042  2005-08-23 22:43:07.000  2005-08-31 21:33:07.000           2005-08-26
16043  2005-08-23 22:50:12.000  2005-08-30 01:01:12.000           2005-08-26

[16044 rows x 3 columns]

# CREATE TABLE and INSERT 

In [1]:
from pathlib import Path
from sqlalchemy import create_engine
import sqlite3
import pandas as pd


Path('lectures.db').touch()
conn = sqlite3.connect('lectures.db')
c = conn.cursor()
c.execute('''CREATE TABLE my_first_table (first_column TEXT,
                                          second_column INTEGER);''')
c.execute('''INSERT INTO my_first_table (first_column, second_column) 
                                 VALUES ('text value', 12);''')

engine = create_engine("sqlite:///lectures.db")
conn.commit() # ในไพธอน ต้องมีบรรทัดนี้ จึงจะอัพเดตข้อมูลได้
pd.read_sql_query('''SELECT * FROM my_first_table;''', engine)

first_column  second_column
0   text value             12

# ARRAY datatype
#### Arrays are not supported as datatype for a column value in SQLite.

## INSERT statements with ARRAYS

In [2]:
c.execute('''CREATE TABLE grades (student_id INT, email TEXT[][], test_scores INT[]);''')
c.execute('''INSERT INTO  grades VALUES (1, 
                                         '{{"work","work1@datacamp.com"},{"other","other1@datacamp.com"}}',
                                          '{92,85,96,88}');''')
conn.commit() 

# Accessing ARRAYs

In [3]:
pd.read_sql_query('''
SELECT * FROM grades''', engine)

student_id                                              email  \
0           1  {{"work","work1@datacamp.com"},{"other","other...   

     test_scores  
0  {92,85,96,88}

```
SELECT email[1][1] AS type,
       email[1][2] AS address,
       test_scores[1]
       FROM grades;''', engine)

+--------+--------------------+-------------+
| type   | address            | test_scores |
|--------|--------------------|-------------|
| work   | work1@datacamp.com | 92          |
| other  | other1@datacamp.com| 85          |
+--------+--------------------+-------------+

```
 
# ARRAY functions and operators

```
SELECT email[2][1] AS type,
       email[2][2] AS address,
       test_scores[1]
FROM grades
WHERE 'other' = ANY (email);

```

```
SELECT email[2][1] AS type,
       email[2][2] AS address,
       test_scores[1]
FROM grades
WHERE email @> ARRAY['other'];
```
These two queries return the same result

```
+---------+---------------------+-------------+
| type    | address             | test_scores |
|---------|-----------------------------------|
| other   | other1@datacamp.com | 92          |
| null    | null                | 76          |
+---------+---------------------+-------------+
```

### Accessing data in an ARRAY

In our DVD Rentals database, the film table contains an ARRAY for `special_features` which has a type of `TEXT[]`. Much like any `ARRAY` data type in PostgreSQL, a `TEXT[]` array can store an array of `TEXT` values. This comes in handy when you want to store things like phone numbers or email addresses as we saw in the lesson.

Let's take a look at the `special_features` column and also practice accessing data in the ARRAY.

* Select the title and special features from the `film` table and compare the results between the two columns.

In [11]:
engine = create_engine("sqlite:///sqlite-sakila.db")
pd.read_sql_query('''SELECT title, special_features FROM film WHERE (title NOT LIKE 'A%') AND (title NOT LIKE 'BA%') ;''', engine).head()

title                      special_features
0  BEACH HEARTBREAKERS      Deleted Scenes,Behind the Scenes
1       BEAR GRACELAND                        Deleted Scenes
2      BEAST HUNCHBACK      Deleted Scenes,Behind the Scenes
3        BEAUTY GREASE                 Trailers,Commentaries
4         BED HIGHBALL  Trailers,Commentaries,Deleted Scenes

* Select all films that have a special feature `Trailers` by filtering on the first index of the `special_features` ARRAY.

```
SELECT title, 
       special_features 
FROM film
WHERE special_features[1] = 'Trailers';
```

```
+---------------------+-------------------------------------------+
| title	              |               special_features            |
-------------------------------------------------------------------
| BEDAZZLED MARRIED   |	Trailers,Deleted Scenes,Behind the Scenes |
| BEHAVIOR RUNAWAY    |	Trailers,Deleted Scenes,Behind the Scenes |
| BLUES INSTINCT	  | Trailers,Deleted Scenes,Behind the Scenes |
+---------------------+-------------------------------------------+
```

* Now let's select all films that have `Deleted Scenes` in the second index of the `special_features` ARRAY.

```
SELECT title, special_features 
FROM film
WHERE special_features[2] = 'Deleted Scenes';
```

```
title	            special_features
BEDAZZLED MARRIED	Trailers,Deleted Scenes,Behind the Scenes
BEHAVIOR RUNAWAY	Trailers,Deleted Scenes,Behind the Scenes
BETRAYED REAR	    Commentaries,Deleted Scenes,Behind the Scenes
```

### Searching an ARRAY with ANY
PostgreSQL also provides the ability to filter results by searching for values in an ARRAY. The `ANY` function allows you to search for a value in any index position of an ARRAY. Here's an example.

```
WHERE 'search text' = ANY(array_name)
```

When using the `ANY` function, the value you are filtering on appears on the left side of the equation with the name of the ARRAY column as the parameter in the `ANY` function.

* Match `'Trailers'` in any index of the `special_features` ARRAY regardless of position.

```
SELECT title, special_features 
FROM film 
WHERE 'Trailers' = ANY(special_features);
```

### Searching an ARRAY with @>
The contains operator `@>` operator is alternative syntax to the `ANY` function and matches data in an ARRAY using the following syntax.

```
WHERE array_name @> ARRAY['search text'] :: type[]
```

* Use the contains operator to match the text `Deleted Scenes` in the `special_features` column.

```
SELECT title, special_features 
FROM film 
WHERE special_features @> ARRAY['Deleted Scenes'];
```

```
title	            special_features
BEACH HEARTBREAKERS	Deleted Scenes,Behind the Scenes
BEAST HUNCHBACK	    Deleted Scenes,Behind the Scenes
BEDAZZLED MARRIED	Trailers,Deleted Scenes,Behind the Scenes
```

# Adding and subtracting date and time values

Calculate the actual number of days rented as well as the true `expected_return_date` by using the `rental_duration` column from the `film` table along with the familiar `rental_date` from the `rental` table.

This will require that you dust off the skills you learned from prior courses on how to join two or more tables together. To select columns from both the `film` and `rental` tables in a single query, we'll need to use the `inventory` table to join these two tables together since there is no explicit relationship between them. 

* Subtract the `rental_date` from the `return_date` to calculate the number of `days_rented`.

In [30]:
pd.read_sql_query('''
SELECT title, rental_date, return_date, rental_duration,
       CAST(ROUND(JULIANDAY(return_date) - JULIANDAY(rental_date)) AS INT) AS days_rented
FROM film  f
INNER JOIN inventory  i USING(film_id)
INNER JOIN rental  r  USING(inventory_id)
ORDER BY title;''', engine).head()

title              rental_date              return_date  \
0  ACADEMY DINOSAUR  2005-05-27 07:03:28.000  2005-05-31 08:01:28.000   
1  ACADEMY DINOSAUR  2005-05-30 20:21:07.000  2005-06-06 00:36:07.000   
2  ACADEMY DINOSAUR  2005-06-15 02:57:51.000  2005-06-20 01:41:51.000   
3  ACADEMY DINOSAUR  2005-06-17 20:24:00.000  2005-06-23 17:45:00.000   
4  ACADEMY DINOSAUR  2005-06-21 00:30:26.000  2005-06-28 03:42:26.000   

   rental_duration  days_rented  
0                6          4.0  
1                6          6.0  
2                6          5.0  
3                6          6.0  
4                6          7.0

* Use the `AGE()` function to calculate the `days_rented`

```
SELECT title, rental_duration,
	   AGE(return_date, rental_date) AS days_rented
FROM film  f
INNER JOIN inventory  i USING(film_id) 
INNER JOIN rental  r USING(inventory_id)
ORDER BY title;
```

query นี้ ให้ผลคล้ายๆกับการเอาคอลัมน์ที่เป็น time stamp มาลบกัน

# INTERVAL arithmetic

If you were running a real DVD Rental store, there would be times when you would need to determine what film titles were currently out for rental with customers. In the previous exercise, we saw that some of the records in the results had a `NULL` value for the `return_date`. This is because the rental was still outstanding.

Each rental in the `film` table has an associated `rental_duration` column which represents the number of days that a DVD can be rented by a customer before it is considered late. In this example, you will exclude films that have a `NULL` value for the `return_date` and also convert the `rental_duration` to an `INTERVAL` type. Here's a reminder of one method for performing this conversion.

```
SELECT INTERVAL '1' day * timestamp '2019-04-10 12:34:56'
```

* Convert `rental_duration` by multiplying it with a 1 day `INTERVAL`
* Subtract the `rental_date` from the `return_date` to calculate the number of `days_rented`.
* Exclude rentals with a `NULL` value for `return_date`.

```
SELECT title,
 	   INTERVAL '1' day * rental_duration,
       return_date - rental_date AS days_rented
FROM film f
INNER JOIN inventory i USING(film_id)
INNER JOIN rental r USING(inventory_id)
-- Filter the query to exclude outstanding rentals
WHERE return_date IS NOT NULL
ORDER BY title;
```

# Calculating the expected return date

So now that you've practiced how to add and subtract timestamps and perform relative calculations using intervals, let's use those new skills to calculate the actual expected return date of a specific rental. As you've seen in previous exercises, the `rental_duration` is the number of days allowed for a rental before it's considered late. To calculate the `expected_return_date` you will want to use the `rental_duration` and add it to the `rental_date`.

* Convert `rental_duration` by multiplying it with a 1-day `INTERVAL`.
* Add it to the rental date.

```
SELECT title,
	   rental_date,
       rental_duration,
       INTERVAL '1' day * rental_duration + rental_date AS expected_return_date,
       return_date
FROM film f
INNER JOIN inventory i  USING(film_id)
INNER JOIN rental r USING(inventory_id)
ORDER BY title;
```

# NOW() and CURRENT TIMESTAMP

* SQLite ไม่มี `NOW()`
* `CURRENT_TIMESTAMP` ใส่ precision ไม่ได้ (ไม่มี `CURRENT_TIMESTAMP(2)` อะไรแบบนั้น)

In [41]:
pd.read_sql("SELECT CURRENT_TIMESTAMP, CURRENT_DATE, CURRENT_TIME;", engine)

CURRENT_TIMESTAMP CURRENT_DATE CURRENT_TIME
0  2021-10-14 17:11:38   2021-10-14     17:11:38

### Manipulating the current date and time

* Select a timestamp five days from now and alias it as `five_days_from_now`.

```
SELECT CURRENT_TIMESTAMP::timestamp AS right_now,
       INTERVAL '5 days' + CURRENT_TIMESTAMP AS five_days_from_now;
```

In [44]:
pd.read_sql("SELECT CURRENT_TIMESTAMP AS right_now, DATE(CURRENT_TIMESTAMP, '+5 days') AS five_days_from_now", engine)

right_now five_days_from_now
0  2021-10-14 17:25:51         2021-10-19

# Extracting and transforming date / time data

* EXTRACT(datetime_value FROM source)
    * `SELECT EXTRACT(quarter FROM timestamp '2005-01-04 05:12:00') AS quarter;`

* DATE_PART('field', source)
    * `SELECT DATE_PART('quarter', timestamp '2005-01-04 05:12:00') AS quarter;`

### Using EXTRACT

You can use `EXTRACT()` and `DATE_PART()` to easily create new fields in your queries by extracting sub-fields from a source timestamp field.

Now suppose you want to produce a predictive model that will help forecast DVD rental activity by day of the week. You could use the `EXTRACT()` function with the `dow` field identifier in query to create a new field called `dayofweek` as a sub-field of the `rental_date` column from the `rental` table.

You can `COUNT()` the number of records in the rental table for a given date range and aggregate by the newly created `dayofweek` column.

* Get the day of the week from the `rental_date` column.

```
SELECT EXTRACT(dow FROM rental_date) AS dayofweek 
FROM rental
```

In [51]:
pd.read_sql_query('''
SELECT STRFTIME('%w', rental_date) AS dayofweek FROM rental;''', engine).head()

dayofweek
0         2
1         2
2         2
3         2
4         2

* Count the total number of rentals by day of the week.

```
SELECT EXTRACT(dow FROM rental_date) AS dayofweek, 
       COUNT(*) as rentals 
FROM rental 
GROUP BY 1;
```

In [53]:
pd.read_sql_query('''
SELECT STRFTIME('%w', rental_date) AS dayofweek, 
       COUNT(*) AS rentals 
FROM rental 
GROUP BY dayofweek;''', engine)

dayofweek  rentals
0         0     2320
1         1     2247
2         2     2463
3         3     2231
4         4     2200
5         5     2272
6         6     2311

### Using DATE_TRUNC

The `DATE_TRUNC()` function will truncate timestamp or interval data types to return a timestamp or interval at a specified precision. The precision values are a subset of the field identifiers that can be used with the `EXTRACT()` and `DATE_PART()` functions. `DATE_TRUNC()` will return an interval or timestamp rather than a number. For example

```
SELECT DATE_TRUNC('month', TIMESTAMP '2005-05-21 15:30:30');
Result: 2005-05-01 00;00:00
```

Now, let's experiment with different precisions and ultimately modify the queries from the previous exercises to aggregate rental activity.

* Truncate the rental_date field by year.

```
SELECT DATE_TRUNC('year', rental_date) AS rental_year
FROM rental;
```

* Now modify the previous query to truncate the `rental_date` by month.

```
SELECT DATE_TRUNC('month', rental_date) AS rental_month
FROM rental;
```
* Let's see what happens when we truncate by day of the month.

```
SELECT DATE_TRUNC('day', rental_date) AS rental_day 
FROM rental;
```

* Finally, count the total number of rentals by `rental_day` and alias it as `rentals`.

```
SELECT DATE_TRUNC('day', rental_date) AS rental_day,
       COUNT(*) AS rentals
FROM rental
GROUP BY rental_day;
```

In [58]:
pd.read_sql_query('''
SELECT STRFTIME('%Y-%m-%d', rental_date) AS rental_day,
      COUNT(*) AS rentals
FROM rental
GROUP BY rental_day;''', engine)

rental_day  rentals
0   2005-05-24        8
1   2005-05-25      137
2   2005-05-26      174
3   2005-05-27      166
4   2005-05-28      196
5   2005-05-29      154
6   2005-05-30      158
7   2005-05-31      163
8   2005-06-14       16
9   2005-06-15      348
10  2005-06-16      324
11  2005-06-17      325
12  2005-06-18      344
13  2005-06-19      348
14  2005-06-20      331
15  2005-06-21      275
16  2005-07-05       27
17  2005-07-06      504
18  2005-07-07      461
19  2005-07-08      512
20  2005-07-09      513
21  2005-07-10      480
22  2005-07-11      461
23  2005-07-12      495
24  2005-07-26       33
25  2005-07-27      649
26  2005-07-28      620
27  2005-07-29      641
28  2005-07-30      634
29  2005-07-31      679
30  2005-08-01      671
31  2005-08-02      643
32  2005-08-16       23
33  2005-08-17      593
34  2005-08-18      621
35  2005-08-19      628
36  2005-08-20      624
37  2005-08-21      659
38  2005-08-22      626
39  2005-08-23      598
40  2006-02-14      182

### Putting it all together

You are going to extract a list of customers and their rental history over 90 days. You will be using the `EXTRACT()`, `DATE_TRUNC()`, and `AGE()` functions that you learned about during this chapter along with some general SQL skills from the prerequisites to extract a data set that could be used to determine what day of the week customers are most likely to rent a DVD and the likelihood that they will return the DVD late.

* Extract the day of the week from the `rental_date` column using the alias `dayofweek`.
* Use an `INTERVAL` in the `WHERE` clause to select records for the 90 day period starting on 5/1/2005.

```
SELECT EXTRACT(dow FROM rental_date) AS dayofweek,
       AGE(return_date, rental_date) AS rental_days
FROM rental r 
WHERE rental_date BETWEEN CAST('2005-05-01' AS TIMESTAMP) AND CAST('2005-05-01' AS TIMESTAMP) + INTERVAL '90 day';
```

In [62]:
pd.read_sql_query('''
SELECT STRFTIME('%d', rental_date) AS dayofweek,
       JULIANDAY(return_date) - JULIANDAY(rental_date) AS rental_days
FROM rental r 
WHERE rental_date BETWEEN DATE('2005-05-01') AND DATE('2005-05-01', '+90 days') ;''', engine)

dayofweek  rental_days
0           24     1.965972
1           24     3.865278
2           24     7.964583
3           24     9.110417
4           24     8.227778
...        ...          ...
8858        29     5.848611
8859        29     5.858333
8860        29     0.930556
8861        29     1.219444
8862        29     2.222917

[8863 rows x 2 columns]

* Finally, use a `CASE` statement and `DATE_TRUNC()` to create a new column called `past_due` which will be `TRUE` if the `rental_days` is greater than the `rental_duration` otherwise, it will be `FALSE`.

```
SELECT first_name || ' ' || last_name AS customer_name,
       title,
       rental_date,
       EXTRACT(dow FROM rental_date) AS dayofweek,
       AGE(return_date, rental_date) AS rental_days,
  -- Use DATE_TRUNC to get days from the AGE function
       CASE WHEN DATE_TRUNC('day', AGE(return_date, rental_date)) > rental_duration * INTERVAL '1' day THEN TRUE ELSE FALSE END AS past_due 
FROM film f 
INNER JOIN inventory i 
USING (film_id) 
INNER JOIN rental r 
USING(inventory_id) 
INNER JOIN customer c 
USING (customer_id) 
WHERE rental_date BETWEEN CAST('2005-05-01' AS DATE) AND CAST('2005-05-01' AS DATE) + INTERVAL '90 day';
```

สามารถทำใน SQLite ได้ดังนี้

* Query ด้านบน มอง `rental_days = 7 days, 23:09:00` เป็นแค่ 7 ซึ่งจาก SQLite `JULIANDAY` มันเป็น `7.96` แล้ว คำตอบแบบของเราควรจะถูกมากกว่า

In [69]:
pd.read_sql_query('''
SELECT first_name || ' ' || last_name AS customer_name,
       title,
       rental_date,
       STRFTIME('%w', rental_date) AS dayofweek,
       JULIANDAY(return_date) - JULIANDAY(rental_date) AS rental_days,
       rental_duration,
  -- Use DATE_TRUNC to get days from the AGE function
       CASE WHEN JULIANDAY(return_date) - JULIANDAY(rental_date) > rental_duration THEN 'true' ELSE 'false' END AS past_due 
FROM film f 
INNER JOIN inventory i 
USING (film_id) 
INNER JOIN rental r 
USING(inventory_id) 
INNER JOIN customer c 
USING (customer_id) 
WHERE rental_date BETWEEN DATE('2005-05-01') AND DATE('2005-05-01', '+90 days') ;''', engine).head()

customer_name            title              rental_date dayofweek  \
0  CHARLOTTE HUNTER  BLANKET BEVERLY  2005-05-24 22:53:30.000         2   
1     TOMMY COLLAZO     FREAKY POCUS  2005-05-24 22:54:33.000         2   
2    MANUEL MURRELL    GRADUATE LORD  2005-05-24 23:03:39.000         2   
3      ANDREW PURDY    LOVE SUICIDES  2005-05-24 23:04:41.000         2   
4    DELORES HANSEN  IDOLS SNATCHERS  2005-05-24 23:05:21.000         2   

   rental_days  rental_duration past_due  
0     1.965972                7    false  
1     3.865278                7    false  
2     7.964583                7     true  
3     9.110417                6     true  
4     8.227778                5     true

# String Concatenation

* จริงๆจากแบบฝึกหัดด้านบน เราได้เห็นตัวอย่างการต่อ string ไปแล้ว


In [72]:
pd.read_sql_query('''SELECT first_name, last_name, first_name || ' ' || last_name AS full_name FROM customer''', engine).head(2)

first_name last_name         full_name
0       MARY     SMITH        MARY SMITH
1   PATRICIA   JOHNSON  PATRICIA JOHNSON

# CONCAT in PostgreSQL (SQLite ไม่มี `CONCAT()` function)

```
SELECT CONCAT(first_name,' ', last_name) AS full_name
FROM customer;
```

# String concat with a non-string

In [78]:
for i in inspector.get_columns('customer'):
    if i["name"] == "customer_id":
        print(i["name"], i["type"])

pd.read_sql_query('''SELECT customer_id || ': '|| first_name || ' '|| last_name AS full_name FROM customer;''', engine).head(2)

customer_id INTEGER


full_name
0        1: MARY SMITH
1  2: PATRICIA JOHNSON

# UPPER LOWER

In [82]:
pd.read_sql_query('''
SELECT UPPER(email), LOWER(email), email FROM customer''', engine).head(3)

UPPER(email)                         LOWER(email)  \
0        MARY.SMITH@SAKILACUSTOMER.ORG        mary.smith@sakilacustomer.org   
1  PATRICIA.JOHNSON@SAKILACUSTOMER.ORG  patricia.johnson@sakilacustomer.org   
2    LINDA.WILLIAMS@SAKILACUSTOMER.ORG    linda.williams@sakilacustomer.org   

                                 email  
0        MARY.SMITH@sakilacustomer.org  
1  PATRICIA.JOHNSON@sakilacustomer.org  
2    LINDA.WILLIAMS@sakilacustomer.org

# INITCAP to Capitalize the string (Not available in Sqlite)

```SELECT INITCAP(title) FROM film```

# REPLACE

In [90]:
pd.read_sql_query('''
SELECT description, 
       REPLACE(description, 'A Astounding', 'An Astounding' ) AS corrected_description
FROM film;''', engine).head()

description  \
0  A Epic Drama of a Feminist And a Mad Scientist...   
1  A Astounding Epistle of a Database Administrat...   
2  A Astounding Reflection of a Lumberjack And a ...   
3  A Fanciful Documentary of a Frisbee And a Lumb...   
4  A Fast-Paced Documentary of a Pastry Chef And ...   

                               corrected_description  
0  A Epic Drama of a Feminist And a Mad Scientist...  
1  An Astounding Epistle of a Database Administra...  
2  An Astounding Reflection of a Lumberjack And a...  
3  A Fanciful Documentary of a Frisbee And a Lumb...  
4  A Fast-Paced Documentary of a Pastry Chef And ...

# REVERSE (not available in SQLite)

```
SELECT title, 
       REVERSE(title)
FROM film;
```

```
+-------------------------------------+
| title            | reverse(title)   |
|-------------------------------------|
| ACADEMY DINOSAUR | RUASONID YMEDACA |
| ACE GOLDFINGER   | REGNIFDLOG ECA   | 
+-------------------------------------+
```

### All string functions in SQLite is at https://www.sqlitetutorial.net/sqlite-string-functions/.

### Concatenating strings

In this exercise and the ones that follow, we are going to derive new fields from columns within the `customer` and `film` tables of the DVD rental database.

We'll start with the `customer` table and create a query to return the customers name and email address formatted such that we could use it as a "To" field in an email script or program. This format will look like the following:

* Concatenate the `first_name` and `last_name` columns separated by a single space followed by `email` surrounded by `<` and `>`.

```
SELECT CONCAT(first_name, ' ', last_name,  ' <', email, '>') AS full_email 
FROM customer
```

In [96]:
pd.read_sql_query('''
SELECT first_name || ' ' || last_name || ' <' || LOWER(email) || '>' AS full_email 
FROM customer;''', engine).head(3)

full_email
0         MARY SMITH <mary.smith@sakilacustomer.org>
1  PATRICIA JOHNSON <patricia.johnson@sakilacusto...
2  LINDA WILLIAMS <linda.williams@sakilacustomer....

### Changing the case of string data

Now you are going to use the `film` and `category` tables to create a new field called `film_category` by concatenating the category `name` with the film's `title`. You will also format the result using functions you learned about in the video to transform the case of the fields you are selecting in the query; for example, the `INITCAP()` function which converts a string to title case.

* Convert the film category `name` to uppercase.
* Convert the first letter of each word in the film's `title` to upper case.
* Concatenate the converted category `name` and film `title` separated by a colon.
* Convert the `description` column to lowercase.

```
SELECT 
  -- Concatenate the category name to coverted to uppercase
  -- to the film title converted to title case
  UPPER(name)  || ': ' || INITCAP(title) AS film_category, 
  -- Convert the description column to lowercase
  LOWER(description) AS description
FROM film f 
INNER JOIN film_category fc USING(film_id) 
INNER JOIN category c USING(category_id);
```

### Replacing string data

Sometimes you will need to make sure that the data you are extracting does not contain any whitespace. There are many different approaches you can take to cleanse and prepare your data for these situations. A common technique is to replace any whitespace with an underscore.

In this example, we are going to practice finding and replacing whitespace characters in the `title` column of the `film` table using the `REPLACE()` function.

* Replace all whitespace with an underscore.

In [99]:
pd.read_sql_query('''SELECT REPLACE(title, ' ', '_') AS title FROM film; ''', engine).head(3)

title
0  ACADEMY_DINOSAUR
1    ACE_GOLDFINGER
2  ADAPTATION_HOLES

# LENGTH

In [103]:
pd.read_sql_query("SELECT title, LENGTH(title) FROM film;", engine).head(3)

title  LENGTH(title)
0  ACADEMY DINOSAUR             16
1    ACE GOLDFINGER             14
2  ADAPTATION HOLES             16

# POSITION  (Use `INSTR` in SQLite)

`SELECT email, POSITION('@' IN email) FROM customer;`

`SELECT email, STRPOS(email, '@') FROM customer;`

In [106]:
pd.read_sql_query('''SELECT email, INSTR(email, '@') FROM customer;''', engine).head(3)

email  INSTR(email, '@')
0        MARY.SMITH@sakilacustomer.org                 11
1  PATRICIA.JOHNSON@sakilacustomer.org                 17
2    LINDA.WILLIAMS@sakilacustomer.org                 15

# LEFT (Use `SUBSTR(col, start, stop)` in SQLite)

`SELECT LEFT(description, 50) FROM film;`

In [111]:
pd.read_sql_query('''SELECT SUBSTR(description, 1, 50) FROM film;''', engine).head(3)

SUBSTR(description, 1, 50)
0  A Epic Drama of a Feminist And a Mad Scientist who
1  A Astounding Epistle of a Database Administrator A
2  A Astounding Reflection of a Lumberjack And a Car

# RIGHT (`SUBSTR(col, -stop)` in SQLite)

`SELECT RIGHT(description, 50) FROM film;`

In [112]:
pd.read_sql_query('''SELECT SUBSTR(description, -50) FROM film;''', engine).head(3)

SUBSTR(description, -50)
0   who must Battle a Teacher in The Canadian Rockies
1  nd a Explorer who must Find a Car in Ancient China
2  Car who must Sink a Lumberjack in A Baloon Factory

# SUBSTRING (Use `SUBSTR` in SQLite)

``` 
SELECT SUBSTRING(description, 10, 50)
FROM film;
```

In [113]:
pd.read_sql_query('''SELECT SUBSTR(description, 10, 50) FROM film;''', engine).head(3)

SUBSTR(description, 10, 50)
0  ama of a Feminist And a Mad Scientist who must Bat
1  ing Epistle of a Database Administrator And a Expl
2  ing Reflection of a Lumberjack And a Car who must

# SUBSTRING along with other functions

```
SELECT SUBSTRING(email FROM 0 FOR POSITION('@' IN email))
FROM customer;
```

### This must be done using `SUBSTR` along with `INSTR` in SQLite

In [117]:
pd.read_sql_query('''
SELECT SUBSTR(email, 0, INSTR(email, '@'))
FROM customer;''', engine).head(3)

SUBSTR(email, 0, INSTR(email, '@'))
0                          MARY.SMITH
1                    PATRICIA.JOHNSON
2                      LINDA.WILLIAMS

#### Another use

```
SELECT SUBSTRING(email FROM POSITION('@' IN email)+1 FOR CHAR_LENGTH(email))
FROM customer;
```

In [119]:
pd.read_sql_query('''
SELECT SUBSTR(email, INSTR(email, '@')+1, LENGTH(email))
FROM customer;''', engine).head(3)

SUBSTR(email, INSTR(email, '@')+1, LENGTH(email))
0                                sakilacustomer.org
1                                sakilacustomer.org
2                                sakilacustomer.org

# Determining the length of strings

Determining the number of characters in a string is something that you will use frequently when working with data in an SQL database. Many situations will require you to find the length of a string stored in your database. For example, 

* you may need to limit the number of characters that are displayed in an application
* you may need to ensure that a column in your dataset contains values that are all the same length 

In this example, we are going to determine the length of the `description` column in the `film` table of the DVD Rental database.

* Select the `title` and `description` columns from the `film` table.
* Find the number of characters in the `description` column with the alias `desc_len`.

In [120]:
pd.read_sql_query('''
SELECT title, description,
       LENGTH(description) AS desc_len
FROM film;''', engine).head()

title  \
0  ACADEMY DINOSAUR   
1    ACE GOLDFINGER   
2  ADAPTATION HOLES   
3  AFFAIR PREJUDICE   
4       AFRICAN EGG   

                                                                                                             description  \
0                       A Epic Drama of a Feminist And a Mad Scientist who must Battle a Teacher in The Canadian Rockies   
1                   A Astounding Epistle of a Database Administrator And a Explorer who must Find a Car in Ancient China   
2                       A Astounding Reflection of a Lumberjack And a Car who must Sink a Lumberjack in A Baloon Factory   
3                           A Fanciful Documentary of a Frisbee And a Lumberjack who must Chase a Monkey in A Shark Tank   
4  A Fast-Paced Documentary of a Pastry Chef And a Dentist who must Pursue a Forensic Psychologist in The Gulf of Mexico   

   desc_len  
0        96  
1       100  
2        96  
3        92  
4       117

### Truncating strings

In the previous exercise, you calculated the length of the `description` column and noticed that the number of characters varied but most of the results were over 75 characters. There will be many times when you need to truncate a text column to a certain length to meet specific criteria for an application. In this exercise, we will practice getting the first 50 characters of the `description` column.

* Select the first 50 characters of the `description` column with the alias `short_desc`

```
SELECT LEFT(description, 50) AS short_desc
FROM film
```

In [121]:
pd.read_sql_query('''
SELECT SUBSTR(description, 1, 50) AS short_desc
FROM film''', engine).head()

short_desc
0  A Epic Drama of a Feminist And a Mad Scientist who
1  A Astounding Epistle of a Database Administrator A
2  A Astounding Reflection of a Lumberjack And a Car 
3  A Fanciful Documentary of a Frisbee And a Lumberja
4  A Fast-Paced Documentary of a Pastry Chef And a De

If you look at the results of the query you'll notice that there are several results where the last word has been cut off because it hit the 50 character limit.

How can you truncate this text at the last full word that is less than 50 characters?

### Extracting substrings from text data

In this exercise, you are going to practice how to extract substrings from text columns. The Sakila database contains the `address` table which stores the street address for all the rental store locations. You need a list of all the street names where the stores are located but the `address` column also contains the street number. You'll use several functions that you've learned about in the video to manipulate the `address` column and return only the street address.

* Extract only the street address without the street number from the `address` column.
* Use functions to determine the starting and ending position parameters.

```
SELECT SUBSTRING(address FROM POSITION(' ' IN address)+1 FOR LENGTH(address))
FROM address;
```

In [123]:
pd.read_sql_query('''
SELECT SUBSTR(address, INSTR(address, ' ')+1, LENGTH(address))
FROM address;''', engine).head()

SUBSTR(address, INSTR(address, ' ')+1, LENGTH(address))
0                                          MySakila Drive
1                                         MySQL Boulevard
2                                          Workhaven Lane
3                                         Lillydale Drive
4                                               Hanoi Way

### Combining functions for string manipulation

In the next example, we are going to break apart the `email` column from the `customer` table into three new derived fields. Parsing a single column into multiple columns can be useful when you need to work with certain subsets of data. Email addresses have embedded information stored in them that can be parsed out to derive additional information about our data. For example, we can use the techniques we learned about in the video to determine how many of our customers use an email from a specific domain.

* Extract the characters to the left of the `@` of the `email` column in the customer table and alias it as `username`.
* Now use `SUBSTRING` to extract the characters after the `@` of the email column and alias the new derived field as `domain`.

```
SELECT LEFT(email, POSITION('@' IN email)-1) AS username,
       SUBSTRING(email FROM POSITION('@' IN email)+1 FOR LENGTH(email)) AS domain
FROM customer;
```

In [129]:
pd.read_sql_query('''
SELECT SUBSTR(email, 1,                   INSTR(email, '@')-1) AS username,
       SUBSTR(email, INSTR(email, '@')+1, LENGTH(email)) AS domain
       
FROM customer;''', engine).head()

username              domain
0        MARY.SMITH  sakilacustomer.org
1  PATRICIA.JOHNSON  sakilacustomer.org
2    LINDA.WILLIAMS  sakilacustomer.org
3     BARBARA.JONES  sakilacustomer.org
4   ELIZABETH.BROWN  sakilacustomer.org

# Removing whitespace using TRIM, LTRIM, RTRIM

In [133]:
pd.read_sql_query("SELECT '   padded   ' AS original, RTRIM('     padded   ') AS trimmed;" , engine)

original      trimmed
0     padded          padded

# Padding

Padding strings is useful in many real-world situations. Earlier in this course, we learned about string concatenation and how to combine the customer's 
* *first and last name separated by a single blank space* and also 
* *combined the customer's full name with their email address*.

The padding functions  are an alternative approach to do this task. To use this approach, you will need to combine and nest functions to determine the length of a string to produce the desired result. Remember when calculating the length of a string you often need to adjust the integer returned to get the proper length or position of a string.

Let's revisit the string concatenation exercise but use padding functions.

* Add a single space to the end or right of the `first_name` column using a padding function.
* Use the `||` operator to concatenate the padded `first_name` to the `last_name` column.

```
SELECT 
	RPAD(first_name, LENGTH(first_name)+1) || last_name AS full_name
FROM customer;
```

ใน SQLite ไม่มีฟังก์ชัน RPAD ดังนั้นจึงต้องใช้ `||`  เหมือนเดิม

In [136]:
pd.read_sql_query('''
SELECT first_name || ' ' || last_name AS full_name FROM customer  ''', engine).head()

full_name
0        MARY SMITH
1  PATRICIA JOHNSON
2    LINDA WILLIAMS
3     BARBARA JONES
4   ELIZABETH BROWN

* Now add a single space to the left or beginning of the `last_name` column using a different padding function than the first step.
* Use the `||` operator to concatenate the `first_name` column to the padded `last_name`.

```
SELECT 
	first_name || LPAD(last_name, LENGTH(last_name)+1) AS full_name
FROM customer; 
```

* Add a single space to the right or end of the `first_name` column.
* Add the characters `<` to the right or end of `last_name` column.
* Finally, add the characters `>` to the right or end of the `email` column.

```
SELECT 
	RPAD(first_name, LENGTH(first_name)+1) 
    || RPAD(last_name, LENGTH(last_name)+2, ' <') 
    || RPAD(email, LENGTH(email)+1, '>') AS full_email
FROM customer;
```

In [137]:
pd.read_sql_query('''
SELECT first_name || ' ' || last_name || ' <' || LOWER(email) || '>' AS full_email 
FROM customer;''', engine).head(3)

full_email
0              MARY SMITH <mary.smith@sakilacustomer.org>
1  PATRICIA JOHNSON <patricia.johnson@sakilacustomer.org>
2      LINDA WILLIAMS <linda.williams@sakilacustomer.org>

# The TRIM function

You used the `LEFT()` function to truncate the `description` column to 50 characters but saw that some words were cut off and/or had trailing whitespace. We can use trimming functions to eliminate the whitespace at the end of the string after it's been truncated.

* Convert the film category `name` to uppercase and use the `CONCAT()` concatenate it with the `title`.
* Truncate the description to the first 50 characters and make sure there is no leading or trailing whitespace after truncating.

```
SELECT CONCAT(UPPER(name), ': ', title) AS film_category, 
       TRIM(LEFT(description, 50)) AS film_desc
FROM film
INNER JOIN film_category USING(film_id) 
INNER JOIN category USING(category_id);
```

In [139]:
pd.read_sql_query('''
SELECT UPPER(name) || ': ' || title AS film_category, 
       TRIM(SUBSTR(description, 1, 50)) AS film_desc
FROM film
INNER JOIN film_category USING(film_id) 
INNER JOIN category USING(category_id);''', engine).head()

film_category  \
0  DOCUMENTARY: ACADEMY DINOSAUR   
1         HORROR: ACE GOLDFINGER   
2  DOCUMENTARY: ADAPTATION HOLES   
3       HORROR: AFFAIR PREJUDICE   
4            FAMILY: AFRICAN EGG   

                                            film_desc  
0  A Epic Drama of a Feminist And a Mad Scientist who  
1  A Astounding Epistle of a Database Administrator A  
2   A Astounding Reflection of a Lumberjack And a Car  
3  A Fanciful Documentary of a Frisbee And a Lumberja  
4  A Fast-Paced Documentary of a Pastry Chef And a De

### Putting it all together

In this exercise, we are going to use the `film` and `category` tables to create a new field called `film_category` by concatenating the category `name` with the film's `title`. You will also practice how to truncate text fields like the `film` table's `description` column <ins>*without cutting off a word*</ins>.

To accomplish this we will use the `REVERSE()` function to help determine the position of the last whitespace character in the `description` before we reach 50 characters. This technique can be used to determine the position of the last character that you want to truncate and ensure that it is less than or equal to 50 characters AND does not cut off a word.

* Get the first 50 characters of the `description` column
* Determine the position of the last whitespace character of the truncated `description` column and subtract it from the number 50 as the second parameter in the first function above.

```
SELECT 
  UPPER(name) || ': ' || title AS film_category, 
  -- Truncate the description without cutting off a word
  LEFT(description, 50 - 
    -- Subtract the position of the first whitespace character
    POSITION(
      ' ' IN REVERSE(LEFT(description, 50))
    )
  ) 
FROM film
INNER JOIN film_category USING(film_id) 
INNER JOIN category USING(category_id);
```

ไม่มี `REVERSE` ใน SQLite และวิธีจำลองก็ยุ่งยากเกินไป

# Full-text search

```
SELECT title, description
FROM film
WHERE to_tsvector(title) @@ to_tsquery('elf')
```

```
+----------------------+
| title                |
+----------------------+
| ELF PARTY            |
| ENCINO ELF           | 
| GHOSTBUSTERS ELF     |
+----------------------+
```

### A review of the LIKE operator

The `LIKE` operator allows us to filter our queries by matching one or more characters in text data. By using the `%` wildcard we can match one or more characters in a string. This is useful when you want to return a result set that matches certain characteristics and can also be very helpful during exploratory data analysis or data cleansing tasks.

Let's explore how different usage of the `%` wildcard will return different results by looking at the `film` table of the Sakila DVD Rental database.

* Select all columns for all records that begin with the word `GOLD`.

In [142]:
pd.read_sql("SELECT * FROM film WHERE title LIKE 'GOLD%';", engine).head(2)

film_id                   title  \
0      365              GOLD RIVER   
1      366  GOLDFINGER SENSIBILITY   

                                                                                                        description  \
0  A Taut Documentary of a Database Administrator And a Waitress who must Reach a Mad Scientist in A Baloon Factory   
1                   A Insightful Drama of a Mad Scientist And a Hunter who must Defeat a Pastry Chef in New Orleans   

  release_year  language_id original_language_id  rental_duration  \
0         2006            1                 None                4   
1         2006            1                 None                3   

   rental_rate  length  replacement_cost rating  \
0         4.99     154             21.99      R   
1         0.99      93             29.99      G   

                                         special_features          last_update  
0  Trailers,Commentaries,Deleted Scenes,Behind the Scenes  2021-03-06 15:52:03  
1                 Trailers,Commentaries,Behind the Scenes  2021-03-06 15:52:03

* Now select all records that end with the word `GOLD`.

In [143]:
pd.read_sql("SELECT * FROM film WHERE title LIKE '%GOLD';", engine).head(2)

film_id       title  \
0      644  OSCAR GOLD   
1      870  SWARM GOLD   

                                                                                              description  \
0        A Insightful Tale of a Database Administrator And a Dog who must Face a Madman in Soviet Georgia   
1  A Insightful Panorama of a Crocodile And a Boat who must Conquer a Sumo Wrestler in A MySQL Convention   

  release_year  language_id original_language_id  rental_duration  \
0         2006            1                 None                7   
1         2006            1                 None                4   

   rental_rate  length  replacement_cost rating       special_features  \
0         2.99     115             29.99     PG      Behind the Scenes   
1         0.99     123             12.99  PG-13  Trailers,Commentaries   

           last_update  
0  2021-03-06 15:52:05  
1  2021-03-06 15:52:07

* Finally, select all records that contain the word 'GOLD'.

In [144]:
pd.read_sql("SELECT * FROM film WHERE title LIKE '%GOLD%';", engine).head(2)

film_id                 title  \
0        2        ACE GOLDFINGER   
1       95  BREAKFAST GOLDFINGER   

                                                                                            description  \
0  A Astounding Epistle of a Database Administrator And a Explorer who must Find a Car in Ancient China   
1                    A Beautiful Reflection of a Student And a Student who must Fight a Moose in Berlin   

  release_year  language_id original_language_id  rental_duration  \
0         2006            1                 None                3   
1         2006            1                 None                5   

   rental_rate  length  replacement_cost rating  \
0         4.99      48             12.99      G   
1         4.99     123             18.99      G   

                       special_features          last_update  
0               Trailers,Deleted Scenes  2021-03-06 15:52:00  
1  Trailers,Commentaries,Deleted Scenes  2021-03-06 15:52:01

# What is a tsvector?

Convert a text column from the `film` table to a `tsvector` and inspect the results. Understanding how full-text search works is the first step in more advanced machine learning and data science concepts like natural language processing.

* Select the film `description` and convert it to a `tsvector` data type.

```
SELECT to_tsvector(description)
FROM film;
```

```
to_tsvector
--------------------
'display':3 'fate':2 'georgia':19 'mad':9 'must':12 'outgun':13 'scientist':10 'shark':16 'soviet':18 'woman':6
'ancient':18 'awe':3 'awe-inspir':2 'boy':16 'china':19 'defeat':14 'epistl':5 'inspir':4 'must':13 'squirrel':11 'student':8
'abandon':19 'astound':2 'charact':3 'fun':20 'hous':21 'mad':15 'madman':7 'meet':13 'must':12 'robot':10 'scientist':16 'studi':4
'berlin':17 'drama':3 'husband':9 'must':11 'outrac':12 'student':6 'sumo':14 'unbeliev':2 'wrestler':15
```

### Basic full-text search

Searching text will become something you do repeatedly when building applications or exploring data sets for data science. Full-text search is helpful when performing exploratory data analysis for a natural language processing model or building a search feature into your application.

In this exercise, you will practice searching a text column and match it against a string. The search will return the same result as a query that uses the `LIKE` operator with the `%` wildcard at the beginning and end of the string, but will perform much better and provide you with a foundation for more advanced full-text search queries. 

* Select the `title` and `description` columns from the `film` table.
* Perform a full-text search on the `title` column for the word `elf`.

```
SELECT title, description
FROM film
WHERE to_tsvector(title) @@  to_tsquery('elf');
```

```
title	                                                        description
-------------------------------------------------------------------------------------------------------------
GHOSTBUSTERS ELF	A Thoughtful Epistle of a Dog And a Feminist who must Chase a Composer in Berlin
ELF MURDER	A Action-Packed Story of a Frisbee And a Woman who must Reach a Girl in An Abandoned Mine Shaft
ENCINO ELF	A Astounding Drama of a Feminist And a Teacher who must Confront a Husband in A Baloon
```